File: 207-course-schedule/course-schedule.cpp
// Brute Force Code & Optimal Code [DFS Method]
class Solution {
public:
    bool isCycleDFS(unordered_map<int, vector<int>>& adj, int vertex, vector<bool>& visited, vector<bool>& inRecursion) {
        // mark current node as visited
        visited[vertex] = true;
        // mark current node as part of the current DFS path
        inRecursion[vertex] = true;
        // visit all neighbours
        for (int neighbour : adj[vertex]) {
            // if neighbour is not visited continue DFS from neighbour
            if (!visited[neighbour]) {
                if (isCycleDFS(adj, neighbour, visited, inRecursion)) {
                    return true;
                }
            }
            // if neighbour is already in the current DFS path then a cycle is found
            else if (inRecursion[neighbour]) {
                return true;
            }
        }
        // DFS of this node is complete remove it from the current DFS path
        inRecursion[vertex] = false;
        // no cycle found from this node
        return false;
    }
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        // adjacency list
        unordered_map<int, vector<int>> adj;
        // visited[i] = true if course i was visited before
        vector<bool> visited(numCourses, false);
        // inRecursion[i] = true if course i is currently present in the DFS path
        vector<bool> inRecursion(numCourses, false);
        // build the directed graph
        for (auto& vec : prerequisites) {
            // "a" = course to be taken
            int a = vec[0];
            // "b" = prerequisite course
            int b = vec[1];
            // create edge: b ---> a
            adj[b].push_back(a);
        }
        // check every course
        for (int i = 0; i < numCourses; i++) {
            // start DFS only if course is not visited
            if (!visited[i]) {
                // if DFS detects a cycle courses cannot be completed
                if (isCycleDFS(adj, i, visited, inRecursion)) {
                    return false;
                }
            }
        }
        // no cycle found, so all courses can be completed
        return true;
    }
};

// Time Complexity : O(N)
// Space Complexity : O(N)

File: 207-course-schedule/course-schedule.py
# Brute Force Code & Optimal Code [BFS Method]
from collections import defaultdict, deque 
class Solution: 
    def topologicalSortCheck(self, adj, n, indegree): 
        # queue for nodes whose indegree becomes 0
        que = deque() 
        # count how many nodes we can process
        count = 0 
        # add all nodes having indegree = 0 these nodes have no pending prerequisites
        for i in range(n): 
            if indegree[i] == 0: 
                count += 1 
                que.append(i) 
        # BFS using Kahn's Algorithm
        while que: 
            # take a course with no remaining prerequisite
            u = que.popleft() 
            # visit all courses that depend on course u
            for v in adj[u]: 
                # one prerequisite of v is now completed
                indegree[v] -= 1 
                # if all prerequisites of v are completed
                if indegree[v] == 0:
                    # v can now be completed
                    count += 1 
                    que.append(v) 
        # if were able to process all courses then there is no cycle
        if count == n: 
            return True 
        # some courses could not be processed which means a cycle is present
        return False 
    def canFinish(self, numCourses, prerequisites): 
        # adjacency list adj[b] contains courses that can be taken after b
        adj = defaultdict(list) 
        # indegree[i] = number of prerequisites for course i
        indegree = [0] * numCourses 
        # process every prerequisite pair
        for vec in prerequisites:
            # 'a' is the course we want to take
            a = vec[0] 
            # 'b' is the prerequisite course
            b = vec[1] 
            # create edge: b ---> a must complete b before taking a
            adj[b].append(a) 
            # one prerequisite is going into course a
            indegree[a] += 1 
        # check whether all courses can be processed if a cycle is present, not possible
        return self.topologicalSortCheck(adj, numCourses, indegree)

# Time Complexity : O(N)
# Space Complexity : O(N)